src/data_hub/dataframe.py

```diff
@@ -0,0 +1,159 @@
+"""
+This module defines a class called DataFrame that is a drop-in
+replacement for a pandas DataFrame, but also allows automatic
+updating of Deep Origin databases.
+"""
+
+from datetime import datetime
+from typing import Optional
+
+import humanize
+import pandas as pd
+from deeporigin.data_hub import api
+from deeporigin.utils import construct_resource_url
+
+
+class DataFrame(pd.DataFrame):
+    """A subclass of pandas DataFrame that allows for automatic updates to Deep Origin databases. This can be used as a drop-in replacement for a pandas DataFrame, and should support all methods a pandas DataFrame supports.
+
+    The primary method of creating an object of this type is to use the [api.get_dataframe][src.data_hub.api.get_dataframe] function.
+    """
+
+    auto_sync: bool = False
+    """When True, changes made to the dataframe will be automatically synced to the Deep Origin database this dataframe represents."""
+
+    class AtIndexer:
+        """this class override is used to intercept calls to at indexer of a pandas dataframe"""
+
+        def __init__(self, obj):
+            self.obj = obj
+
+        def __getitem__(self, key):
+            """intercept for the set operation"""
+
+            return self.obj._get_value(*key)
+
+        def __setitem__(self, key, value):
+            """intercept for the set operation""" ""
+
+            old_value = self.obj._get_value(*key)
+            if value == old_value:
+                # noop
+                return
+
+            rows = [key[0]]
+            columns = [key[1]]
+
+            # Perform the actual setting operation
+            self.obj._set_value(*key, value)
+
+            # now update the DB. note that self is an AtIndexer
+            # object, so we need to index into the pandas object
+            if self.obj.auto_sync:
+                self.obj.sync(columns=columns, rows=rows)
+
+    @property
+    def at(self):
+        """Override the `at` property to return an AtIndexer"""
+        return self.AtIndexer(self)
+
+    def __setitem__(self, key, value):
```

Comment:
What about intercepting item deletion?

Reply:
As in, deleting a whole column? Good point. Haven't thought about it.

Reply:
I don't think its a priority to support deleting columns now. Instead, we could more simply document what is and isn't supported, and leave the rest to later.